### PR TITLE
Create test-runner command for rerunning tests

### PR DESCRIPTION
**Pull request by bergundy**
## What changed?

Added a test-runner command for rerunning tests and modified the Makefile to use the new command.

## Why?

gotestsum is supposed to do this but it is buggy: https://github.com/gotestyourself/gotestsum/issues/423.

If this runner works well we can stop using gotestsum if we choose but I would probably recommend using it since it claims to do exactly what we need. 
An alternative would be to either generate test reports using JSON output or using https://github.com/jstemmer/go-junit-report to create junit XML reports.

## How did you test it?

- Ran functional tests multiple times locally
- Added unit tests to the runner

### CODE REVIEW

File: tests/callbacks_test.go [hunk 1]
```diff
@@ -96,6 +96,7 @@ func (s *CallbacksSuite) runNexusCompletionHTTPServer(h *completionHandler, list
 }
 
 func (s *CallbacksSuite) TestWorkflowCallbacks_InvalidArgument() {
+	s.Equal(1, 2)
```
Comment (bergundy) [inline tests/callbacks_test.go:99]:
```suggestion
```

File: tools/testrunner/testrunner.go [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+	number       int
+	junitXmlPath string
+	exitErr      *exec.ExitError
+	suites       junit.Testsuites
+}
+
+func (a *attempt) run(ctx context.Context, args []string) error {
+	args = append([]string{"--junitfile", a.junitXmlPath}, args...)
+	log.Printf("starting test attempt %d with args: %v", a.number, args)
+	// This must run from the repo root.
```
Comment (bergundy) [inline tools/testrunner/testrunner.go:53]:
```suggestion
```

File: Makefile [hunk 3]
```diff
@@ -405,13 +402,13 @@ prepare-coverage-test: $(GOTESTSUM) $(TEST_OUTPUT_ROOT)
 
 unit-test-coverage: prepare-coverage-test
 	@printf $(COLOR) "Run unit tests with coverage..."
-	$(GOTESTSUM) --rerun-fails=$(FAILED_TEST_RETRIES) --rerun-fails-max-failures=10 --junitfile $(NEW_REPORT) --packages $(UNIT_TEST_DIRS) -- \
+	go run ./cmd/tools/test-runner $(GOTESTSUM) -retries $(FAILED_TEST_RETRIES) --junitfile $(NEW_REPORT) --packages $(UNIT_TEST_DIRS) -- \
```
Comment (bergundy) [inline Makefile:405]: I think there's a way to pre-build this in our CI setup. I can look into that if anyone thinks using `go run` would be an issue.

Comment (dnr) [reply Makefile:405]: Anything you build in `pre-build-functional-test-coverage` will go into the per-run cache. That won't cover the final binary (I think) but the object files. I wouldn't worry about it, though, the compile+link time should be negligible.

File: tools/testrunner/testrunner.go [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+	a.suites = suites
+	return nil
+}
+
+func (a attempt) failures() []string {
+	root := node{children: make(map[string]node)}
+
+	for _, suite := range a.suites.Suites {
+		for _, tc := range suite.Testcases {
+			if tc.Failure != nil {
```
Comment (ychebotarev) [inline tools/testrunner/testrunner.go:86]: if tc.Failure == nil { continue }
less nestings

Comment (bergundy) [reply tools/testrunner/testrunner.go]: Done.

File: tools/testrunner/testrunner.go [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+
+	for _, suite := range a.suites.Suites {
+		if suite.Failures == 0 {
+			continue
+		}
+		for _, tc := range suite.Testcases {
+			if tc.Failure == nil {
+				continue
+			}
+			n := root
```
Comment (ychebotarev) [inline tools/testrunner/testrunner.go:114]: it took some time to realize that 'n' is something like 'node'.
I would call it treeNode (but I in general not a big fun of short names besides "this" substitutes)

File: tools/testrunner/testrunner.go [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+					n.children[part] = child
+				}
+				n = child
+			}
+		}
+	}
+
+	leafFailures := make([]string, 0)
+
+	for path, n := range root.walk() {
```
Comment (ychebotarev) [inline tools/testrunner/testrunner.go:128]: this worth a comment - it took me some time to understand what you are doing and why.
As I understand if test "a/b/c" failed, there will be failures for "a", "a/b", and "a/b/c" in suite.Testcases.
You only want to get "a/b/c". Thus you build a tree, and then collect only leaf nodes, and while walking you build a "path" to that node.

File: tools/testrunner/testrunner.go [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+}
+
+func newRunner(gotestsumExecutable string) *runner {
+	return &runner{
+		gotestsumExecutable: gotestsumExecutable,
+		attempts:            make([]*attempt, 0),
+	}
+}
+
+func (r *runner) sanitizeAndParseArgs(args []string) ([]string, error) {
```
Comment (ychebotarev) [inline tools/testrunner/testrunner.go:152]: I'm surprised there is no arg parser in go stdlib :)

File: tools/testrunner/testrunner.go [hunk 8]
```diff
@@ -0,0 +1,348 @@
...
+		if i == 0 {
+			combined.XMLName = attempt.suites.XMLName
+			combined.Name = attempt.suites.Name
+			combined.Suites = attempt.suites.Suites
+			continue
+		}
+
+		for _, suite := range attempt.suites.Suites {
+			cpy := suite
+			cpy.Name += fmt.Sprintf(" (retry %d)", i)
```
Comment (ychebotarev) [inline tools/testrunner/testrunner.go:244]: will it be (retry 0)?

Comment (ychebotarev) [reply tools/testrunner/testrunner.go:244]: I think I would rather skip it for the first retry

**Review by ychebotarev — APPROVED** [review]

File: tools/testrunner/testrunner.go (hunk 6 above)
Comment (bergundy) [reply tools/testrunner/testrunner.go:128]: Will do, thanks.

File: tools/testrunner/testrunner.go (hunk 7 above)
Comment (bergundy) [reply tools/testrunner/testrunner.go:152]: There is, it's called `flag` but we need to sanitize and replace args as well, and that's not supported. We also cross the `--` delimiter so I ended up just writing this whole custom function.

File: tools/testrunner/testrunner.go (hunk 8 above)
Comment (bergundy) [reply tools/testrunner/testrunner.go:244]: A few lines up we have an `if i == 0 { ... continue }`

File: tools/testrunner/testrunner.go (hunk 5 above)
Comment (alexshtin) [reply tools/testrunner/testrunner.go:114]: go rule: the less is the scope the shorter the name.

File: tools/testrunner/testrunner.go (hunk 7 above)
Comment (alexshtin) [reply tools/testrunner/testrunner.go:152]: there is `flag.Parse()`. I don't know why Roey doesn't want to use it.

File: tools/testrunner/testrunner.go (hunk 6 above)
Comment (alexshtin) [reply tools/testrunner/testrunner.go:128]: modern go features! nice!

**Review by alexshtin — APPROVED** [review]

**Pull request merged**